### PR TITLE
Input batch sharding strategy BATCH

### PR DESCRIPTION
**Pull request by apoorvtintin**
Axlearn currently supports DataPartitionType strategies FULL and REPLICATED for input batches. This PR adds support for a third sharding strategy DataPartitionType.BATCH. 

This sharding strategy shards the input batch on sharding axis inferred as "batch_axis". This ensures batches are replicated on the "model" dimension and thus avoids unnecessary collectives to reshard input batches when Tensor parallelism is used.

### CODE REVIEW

File: axlearn/common/utils.py [hunk 1]
```diff
@@ -607,6 +610,7 @@ def host_to_global_device_array(
     host_arrays: Nested[Union[np.ndarray, Tensor]],
     *,
     partition: DataPartitionType = DataPartitionType.FULL,
```
Comment (ruomingp) [inline axlearn/common/utils.py:612]: I think @markblee plans to remove the `DataPartitionType` enum and rely on https://jax.readthedocs.io/en/latest/_autosummary/jax.make_array_from_process_local_data.html to support flexible partition specs.

**Review by ruomingp — CHANGES_REQUESTED** [review]

File: axlearn/common/utils.py (hunk 1 above)
Comment (apoorvtintin) [reply axlearn/common/utils.py:612]: Thanks that sounds promising. 

Hello @markblee, let me know if this PR is needed till you make your changes, or if you have your design in mind I can reshape the PR to make it compatible with your design.

Comment (hanzhi713) [reply axlearn/common/utils.py:612]: I think `make_array_from_process_local_data` is already merged: https://github.com/apple/axlearn/pull/781

**Comment by hanzhi713** [comment]
@apoorvtintin With this new sharding strategy, how do you plan to support model parallelism across hosts (i.e. per host bs < 1)? If this is not a supported use case, then this change LGTM.  I could imagine the support for model parallelism across hosts is not needed in the short term for `neuron` since the number of chips per host is large.

File: axlearn/experiments/text/gpt/fuji.py [hunk 2]
```diff
@@ -423,6 +423,7 @@ def get_trainer_kwargs(
         raise NotImplementedError(f"Unknown model size {model_size}.")
     model_kwargs = trainer_kwargs.pop("model_kwargs")
     model_kwargs.setdefault("vocab_size", vocab_size)
+    trainer_kwargs["input_partition_type"] = None if backend != "neuron" else DataPartitionType.BATCH
```
Comment (kelvin-zou) [inline axlearn/experiments/text/gpt/fuji.py:426]: Please go with a configmodifier instead of hardcode in the code, otherwise it becomes hard for people to debug

File: axlearn/common/evaler.py [hunk 3]
```diff
@@ -188,11 +194,11 @@ def _pjit(self, fn: Callable) -> Callable:
             in_shardings=(
                 self._model_param_partition_specs,  # model_params.
                 None,  # replicated_inputs (e.g., prng_key).
-                utils.input_partition_spec(),  # per_example_inputs.
+                utils.data_partition_type_to_spec(partition=self.config.input_partition_type, batch_axis_names=self.config.batch_axis_names),  # per_example_inputs.
             ),
             out_shardings=dict(
                 replicated=None,
-                per_example=utils.input_partition_spec(),
+                per_example=utils.data_partition_type_to_spec( partition=self.config.input_partition_type, batch_axis_names=self.config.batch_axis_names),
```
Comment (kelvin-zou) [inline axlearn/common/evaler.py:201]: Did you run through pylint? this line seems quite long.

File: axlearn/common/utils.py [hunk 4]
```diff
@@ -591,14 +591,17 @@ class DataPartitionType(Enum):
...
-
-def data_partition_type_to_spec(partition: DataPartitionType) -> PartitionSpec:
+def data_partition_type_to_spec(partition: DataPartitionType, * , batch_axis_names: Union[str, Sequence[str]] = ("data", "fsdp")) -> PartitionSpec:
     """Returns a PartitionSpec for the given partition type."""
     if partition == DataPartitionType.FULL:
         return input_partition_spec()
     elif partition == DataPartitionType.REPLICATED:
         return None
+    elif partition == DataPartitionType.BATCH:
+        return PartitionSpec(batch_axis_names)
```
Comment (kelvin-zou) [inline axlearn/common/utils.py:604]: Instead of directly assigning PartitionSpec, maybe extend `input_partition_spec ` with an argument batch_axis_names, with default None?

File: axlearn/common/host_array_test.py [hunk 5]
```diff
@@ -37,16 +36,15 @@ def is_supported(
         )
     )
 
-
```
Comment (kelvin-zou) [inline axlearn/common/host_array_test.py:40]: Are you using a different pylint style? Those lines should not be removed.

**Comment by kelvin-zou** [comment]
Sorry for the late review, I should've checked this earlier. 
@apoorvtintin can you actually look into logical batch and use that instead? From what I see here, this feature is a bit redundant (and less flexible as well) than the logical batch size, and with that, probably you don't need this PR if you just want to run something like batch_size=0.5 or 0.25 per chip with model parallelism.

**Comment by hanzhi713** [comment]
> Sorry for the late review, I should've checked this earlier. @apoorvtintin can you actually look into logical batch and use that instead (
> 
> https://github.com/apple/axlearn/blob/2d1fb295801e954d72400a4c69e066254ff0c84d/axlearn/common/input_dispatch.py#L24
> 
> )? From what I see here, this feature is a bit redundant (and less flexible as well) than the logical batch size, and with that, probably you don't need this PR if you just want to run something like batch_size=0.5 or 0.25 per chip with model parallelism.

I think the primary purpose of this PR (as stated in the description) is to avoid the resharding cost from physical bs to logical bs when there's no need to shard data along model axis. This could remove the collective for inputs re-sharding and thus reduce step time. 

@apoorvtintin May I know how much time do we save by removing this collective? If the benefit is not significant, we can stick to the old code.

**Comment by kelvin-zou** [comment]
> I think the primary purpose of this PR (as stated in the description) is to avoid the resharding cost from physical bs to logical bs when there's no need to shard data along model axis. This could remove the collective for inputs re-sharding and thus reduce step time.

But input collective is only triggered within host, which should be insignificant?

**Comment by apoorvtintin** [comment]
> @apoorvtintin With this new sharding strategy, how do you plan to support model parallelism across hosts (i.e. per host bs < 1)? If this is not a supported use case, then this change LGTM. I could imagine the support for model parallelism across hosts is not needed in the short term for `neuron` since the number of chips per host is large.

@hanzhi713 Tensor parallelism is usually used within a device. But yes theoretically tensor parallelism can be across hosts with per host bs<1. To answer your question, yes this PR does not support that case.

**Comment by apoorvtintin** [comment]
> Sorry for the late review, I should've checked this earlier. @apoorvtintin can you actually look into logical batch and use that instead (
> 
> https://github.com/apple/axlearn/blob/2d1fb295801e954d72400a4c69e066254ff0c84d/axlearn/common/input_dispatch.py#L24
> )? From what I see here, this feature is a bit redundant (and less flexible as well) than the logical batch size, and with that, probably you don't need this PR if you just want to run something like batch_size=0.5 or 0.25 per chip with model parallelism.

@kelvin-zou I checked out fractional batches (logical to physical batches). This feature pads '0' batches to get to at least 1 batch per device. Please correct me if I did not understand it correctly.

For sequence parallelism we need all model axis ranks in a group to have replicated batches. Yes theoretically it means we have only 0.25 batches per device (when fsdp=16 and model=4 ==> 16 batches). but the batches are replicated along the model axis. If I use fractional batches with set batches per device to 0.25 the I would get 16 batches but the rest 48 will be padded 0s.

**Comment by apoorvtintin** [comment]
> > Sorry for the late review, I should've checked this earlier. @apoorvtintin can you actually look into logical batch and use that instead (
> > https://github.com/apple/axlearn/blob/2d1fb295801e954d72400a4c69e066254ff0c84d/axlearn/common/input_dispatch.py#L24
> > 
> > )? From what I see here, this feature is a bit redundant (and less flexible as well) than the logical batch size, and with that, probably you don't need this PR if you just want to run something like batch_size=0.5 or 0.25 per chip with model parallelism.
> 
> I think the primary purpose of this PR (as stated in the description) is to avoid the resharding cost from physical bs to logical bs when there's no need to shard data along model axis. This could remove the collective for inputs re-sharding and thus reduce step time.

Correct me if I am wrong, Axlearn currently does not allow input batches to be sharded along the sequence dimension which can be sharded along the model axis. Which means without this PR the only choices are:
- Fractional batches: May not be enough as described in my [previous reply](https://github.com/apple/axlearn/pull/884#issuecomment-2591751866).
- Sharding REPLICATED: Not feasible as it uses too much memory.
- Sharding FULL: We shard the batch axis along all axis available including model. Which means GBS == global_device_count. For optimal sequence parallel training we want GBS == FSDP*DP ranks.

Which brings us to this Sharding BATCH which is implemented by this PR. This allows us to have GBS == FSDP*DP ranks where batches are replicated across model axis.

> @apoorvtintin May I know how much time do we save by removing this collective? If the benefit is not significant, we can stick to the old code.
That depends on the sharding strategy chosen from the options above, if Sharding FULL is chosen then each model parallel group will have to run a reshard to sync batches (like a TP all gather). This example glosses over obvious problems with full sharding like higher batch size (4 times more than we can fit since model=4).

**Comment by apoorvtintin** [comment]
> But input collective is only triggered within host, which should be insignificant?

@kelvin-zou I assume your question assumes we use SHARDING FULL, which has a batch size problem even though we could theoretically achieve batches replicated along 'model' with an extra collective.

File: axlearn/common/utils.py (hunk 1 above)
Comment (markblee) [reply axlearn/common/utils.py:612]: Yes, support for `make_array_from_process_local_data` has been merged. There doesn't seem to be many remaining advantages of passing both a partition type and aux information like batch axis names; instead, we can take partition specs directly:
```
DataPartitionType.FULL -> input_partition_spec()
DataPartitionType.REPLICATED -> None
DataPartitionType.BATCH -> PartitionSpec(batch_axis_names)
```
This brings the API closer to jax native, but may require more changes internally.

File: axlearn/common/utils_test.py [hunk 6]
```diff
@@ -1701,6 +1702,31 @@ def test_length(self):
 class HostToGlobalArrayTest(TestCase):
     """Tests host_to_global_device_array."""
 
+    @pytest.mark.neuron
+    def test_partition_batch(self):
```
Comment (markblee) [inline axlearn/common/utils_test.py:1706]: Are we able to pass the `test_every_other_process` only relying on `make_array_from_process_local_data`? If not, then it seems that input dispatch may still be needed.

**Review by markblee — COMMENTED** [review]
Sorry for the late review.

**Comment by hanzhi713** [comment]
> > @apoorvtintin With this new sharding strategy, how do you plan to support model parallelism across hosts (i.e. per host bs < 1)? If this is not a supported use case, then this change LGTM. I could imagine the support for model parallelism across hosts is not needed in the short term for `neuron` since the number of chips per host is large.
> 
> @hanzhi713 Tensor parallelism is usually used within a device. But yes theoretically tensor parallelism can be across hosts with per host bs<1. To answer your question, yes this PR does not support that case.

With fractional batches (logical batch), yes at first batches are padded to per device bs=1, but then it will be resharded to what the actual per device bs is here: https://github.com/apple/axlearn/blob/141239c3376638d551e10ff1094e6947e6a75f75/axlearn/common/trainer.py#L1139C38-L1139C59

This dispatch function will reshard physical batches to logical batches, e.g. ("data", "fsdp", "model") -> ("data", "fsdp") through an einsum. After this point, batches should be replicated along model parallel axis as you would expect.

**Comment by apoorvtintin** [comment]
Closing in favor of fractional batches
